src/optimizer/query_rewrite.c

```diff
@@ -3499,6 +3523,874 @@ qo_reduce_outer_joined_tables (PARSER_CONTEXT * parser, PT_NODE * spec, PT_NODE
+    case PT_JOIN_LEFT_OUTER:
+      /* fallthrough */
+    case PT_JOIN_RIGHT_OUTER:
+      /* fallthrough */
+    case PT_JOIN_FULL_OUTER:
+      /* fallthrough */
+    case PT_JOIN_UNION:
+      /* fallthrough */
+    default:
+      goto exit_on_fail_with_exclude;
+    }
+
+  PT_SPEC_GET_DB_OBJECT (curr_pk_spec, curr_pk_mop);
+  if (curr_pk_mop == NULL)
+    {
+      /* CTEs and derived tables are excluded. */
+      goto exit_on_fail_with_exclude;
+    }
+
+  for (curr_pk_cons = sm_class_constraints (curr_pk_mop); curr_pk_cons != NULL; curr_pk_cons = curr_pk_cons->next)
+    {
+      if (curr_pk_cons->type == SM_CONSTRAINT_PRIMARY_KEY)
+	{
+	  /* found */
+	  break;
+	}
+    }
+
+  if (curr_pk_cons == NULL || curr_pk_cons->fk_info == NULL)
+    {
+      /* No primary key or no foreign key referencing the primary key. */
+      goto exit_on_fail_with_exclude;
+    }
+
+  /* There must be no non-join predicates. */
+  for (curr_pred = query->info.query.q.select.where; curr_pred != NULL; curr_pred = curr_pred->next)
+    {
+      memset (&info, 0, sizeof (SPEC_CNT_INFO));
+      info.spec = curr_pk_spec;
+
+      next_pred = curr_pred->next;
+      curr_pred->next = NULL;
+
+      parser_walk_tree (parser, curr_pred, qo_get_name_cnt_by_spec, &info, NULL, NULL);
+
+      curr_pred->next = next_pred;
+
+      if (info.my_spec_cnt >= 1)
+	{
+	  if (curr_pred->or_next == NULL && PT_NODE_IS_EXPR (curr_pred) && PT_EXPR_OP (curr_pred) == PT_EQ
+	      && pt_is_attr (PT_EXPR_ARG1 (curr_pred)) && pt_is_attr (PT_EXPR_ARG2 (curr_pred)))
+	    {
+	      reduce_pred_point_list = parser_append_node (pt_point (parser, curr_pred), reduce_pred_point_list);
+	    }
+	  else
+	    {
+	      /* Non-join predicates exist. */
+	      goto exit_on_fail_with_exclude;
+	    }
+	}
+    }
+
+  if (reduce_pred_point_list == NULL)
+    {
+      /* There are no join predicates. */
+      goto exit_on_fail_with_exclude;
+    }
+
+  /* The columns of join predicates must be in the primary key. */
+  for (curr_pred_point = reduce_pred_point_list; curr_pred_point != NULL; curr_pred_point = curr_pred_point->next)
+    {
+      SM_ATTRIBUTE *pk_cons_attr = NULL;
+      PT_NODE *pk_pred_attr = NULL;
+      PT_NODE *arg1 = NULL, *arg2 = NULL;
+      int i = 0;
+
+      curr_pred = curr_pred_point;
+      CAST_POINTER_TO_NODE (curr_pred);
+
+      assert (PT_NODE_IS_EXPR (curr_pred));
+      assert (PT_EXPR_OP (curr_pred) == PT_EQ);
+      assert (PT_NODE_IS_NAME (PT_EXPR_ARG1 (curr_pred)));
+      assert (PT_NODE_IS_NAME (PT_EXPR_ARG2 (curr_pred)));
+
+      arg1 = PT_EXPR_ARG1 (curr_pred);
+      arg2 = PT_EXPR_ARG2 (curr_pred);
+
+      if (PT_SPEC_ID (curr_pk_spec) == PT_NAME_SPEC_ID (arg1))
+	{
+	  pk_pred_attr = PT_EXPR_ARG1 (curr_pred);
+	}
+      else if (PT_SPEC_ID (curr_pk_spec) == PT_NAME_SPEC_ID (arg2))
+	{
+	  pk_pred_attr = PT_EXPR_ARG2 (curr_pred);
+	}
+      else
+	{
+	  /* Already checked before. */
+	  assert (false);
+	  ERROR_SET_ERROR_ONLY (ER_FAILED);
```

Comment:
It seems to be a normal case to me if already checked. could you elaborate on it?

Reply:
This is an impossible case.
And even if we do not set an error, there is no problem if the spec is skipped.
The reason for setting an error and exiting is to reduce inefficiency.

So I'll change it to check only in debug mode using assert().